### PR TITLE
Add Outputting to CabanaPF Framekwork and Rework PFHub Executable

### PR DESCRIPTION
**Pull request by DavidJoy8**
Blocked by #15 

For the below notes, I will use the following definitions: A **codeuser** is someone who uses CabanaPF to write code.  (The PFHub1a class and executable are us serving as a codeuser.)  An **enduser** is someone who runs executables produced by a codeuser.  They don't write/modify code, but use command line options (or config files if the enduser adds those).

- Added two new concepts to the framework: major_output and minor_output.  Codeusers implement these in their derived class
  - For PFHub1a, minor_output is a console print of the free energy and major output is saving the c grid
- Added an add_output function to the framework.  This tells the framework to run one of the output functions at some simulation time.  When the timestep closest to that simulation time is ran, the output function will be run as well.
  - It is not required to register the times in any order.  The framework internally uses a min heap to keep track of the next output time.
  - Adding output can be done partway through the run.  If outputting at the corresponding timestep would require going back in time, an exception is raised instead.
- Added a command line parser.  This lowers the amount of work codeusers need to do and provides a consistent experience to endusers.
  - There is verification that is enabled by default, but can be disabled by a codeuser if they want to do custom defaults or logic
  - 3 required arguments (if verification enabled): grid, dt, and endtime
  - 4 optional arguments: startoutput, endoutput, majoroutputs, minoroutputs
  - 2 flags: log and outputatzero
- The command line parser can add the appropriate outputs to a Runner
- The PFHub1a executable now uses the command line parser and the framework outputting.  The enduser can run their selected version (2017, 2023, or custom) with a specified number (which can be 0) of free energy and grid outputs, either log or linearlly spaced, up to a specified simulation time.

### CODE REVIEW

File: unit_test/test.cpp [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 #include <Cabana_Core.hpp>
 #include <Cabana_Grid.hpp>
+#include <CommandLine.hpp>
```
Comment (streeve) [inline unit_test/test.cpp:5]: I think it's worth splitting these tests into a separate file (and renaming the existing test file) - maybe `CabanaPF_Test_PFHub` and `CabanaPF_Test_CommandLine`?

**Review by streeve — CHANGES_REQUESTED** [review]
This looks great - just looking to separate the test files

File: src/CommandLine.hpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+#define COMMANDLINE_H
+
+#include <Runner.hpp>
+#include <getopt.h>
+
+namespace CabanaPF {
+
+struct CommandLineInput {
+    inline static const std::string USAGE =
+        " --grid <grid points> --dt <timestep size> --endtime <end time> [--majoroutputs <count>] [--minoroutputs "
```
Comment (streeve) [inline src/CommandLine.hpp:11]: A longer help string would also be helpful, e.g. 
```
[--majoroutputs <count>]
    Optional frequency with which to write simulation state to file
```

**Review by streeve — APPROVED** [review]

**Review by stvdwtt — APPROVED** [review]
Looks good to me!

**Pull request merged**